File: java/eyes.ufg.java/src/main/java/com/applitools/eyes/visualgrid/model/RenderBrowserInfo.java
package com.applitools.eyes.visualgrid.model;

import com.applitools.eyes.RectangleSize;
import com.applitools.eyes.selenium.BrowserType;

public class RenderBrowserInfo implements IRenderingBrowserInfo {

    private RectangleSize viewportSize;
    private BrowserType browserType;
    private String platform = "linux";
    private EmulationBaseInfo emulationInfo;
    private IosDeviceInfo iosDeviceInfo;
    private final String sizeMode = "full-page";
    private String baselineEnvName;
    private AndroidDeviceInfo androidDeviceInfo;

    /** internal */
    private String deviceVersion;

    public RenderBrowserInfo(RectangleSize viewportSize, BrowserType browserType, String baselineEnvName) {
        this.viewportSize = viewportSize;
        this.browserType = browserType;
        this.baselineEnvName = baselineEnvName;
        this.platform = getPlatform();
    }

    public RenderBrowserInfo(RectangleSize viewportSize, BrowserType browserType) {
        this.viewportSize = viewportSize;
        this.browserType = browserType;
        this.platform = getPlatform();
    }

    public RenderBrowserInfo(EmulationBaseInfo emulationInfo, String baselineEnvName) {
        this.emulationInfo = emulationInfo;
        this.baselineEnvName = baselineEnvName;
        this.browserType = BrowserType.CHROME;
        this.platform = getPlatform();
    }

    public RenderBrowserInfo(int width, int height) {
        this(new RectangleSize(width, height), BrowserType.CHROME, null);
    }

    public RenderBrowserInfo(EmulationBaseInfo emulationInfo) {
        this.emulationInfo = emulationInfo;
        this.browserType = BrowserType.CHROME;
        this.platform = getPlatform();
    }

    public RenderBrowserInfo(int width, int height, BrowserType browserType, String baselineEnvName) {
        this(new RectangleSize(width, height), browserType, baselineEnvName);
    }

    public RenderBrowserInfo(int width, int height, BrowserType browserType) {
        this(new RectangleSize(width, height), browserType, null);
    }

    public RenderBrowserInfo(IosDeviceInfo deviceInfo) {
        this.iosDeviceInfo = deviceInfo;
        this.browserType = BrowserType.SAFARI;
        this.platform = getPlatform();
    }

    public RenderBrowserInfo(String version, IosDeviceInfo deviceInfo) {
        this.iosDeviceInfo = deviceInfo;
        this.deviceVersion = version;
        this.browserType = BrowserType.SAFARI;
        this.platform = getPlatform();
    }

    public RenderBrowserInfo(IosDeviceInfo deviceInfo, String baselineEnvName) {
        this.iosDeviceInfo = deviceInfo;
        this.baselineEnvName = baselineEnvName;
        this.browserType = BrowserType.SAFARI;
        this.platform = getPlatform();
    }

    public RenderBrowserInfo(AndroidDeviceInfo deviceInfo) {
        this.androidDeviceInfo = deviceInfo;
        this.browserType = BrowserType.CHROME;
        this.platform = getPlatform();
    }

    public RenderBrowserInfo(String version, AndroidDeviceInfo deviceInfo) {
        this.androidDeviceInfo = deviceInfo;
        this.deviceVersion = version;
        this.browserType = BrowserType.CHROME;
        this.platform = getPlatform();
    }

    public RenderBrowserInfo(AndroidDeviceInfo deviceInfo, String baselineEnvName) {
        this.androidDeviceInfo = deviceInfo;
        this.baselineEnvName = baselineEnvName;
        this.browserType = BrowserType.CHROME;
        this.platform = getPlatform();
    }

    public RectangleSize getDeviceSize() {
        if (viewportSize != null) {
            return viewportSize;
        }

        if (emulationInfo != null && emulationInfo.size != null) {
            return emulationInfo.size;
        }

        if (iosDeviceInfo != null && iosDeviceInfo.size != null) {
            return iosDeviceInfo.size;
        }

        if (androidDeviceInfo != null && androidDeviceInfo.size != null) {
            return androidDeviceInfo.size;
        }

        return RectangleSize.EMPTY;
    }

    public void setEmulationDeviceSize(DeviceSize size) {
        if (size != null && emulationInfo != null) {
            if (emulationInfo.screenOrientation.equals(ScreenOrientation.PORTRAIT)) {
                emulationInfo.size = size.getPortrait();
            } else {
                emulationInfo.size = size.getLandscape();
            }
        }
    }

    public void setIosDeviceSize(DeviceSize size) {
        if (size != null && iosDeviceInfo != null) {
            if (iosDeviceInfo.getScreenOrientation().equals(ScreenOrientation.PORTRAIT)) {
                iosDeviceInfo.size = size.getPortrait();
            } else {
                iosDeviceInfo.size = size.getLandscape();
            }
        }
    }

    public int getWidth() {
        if (viewportSize != null) {
            return viewportSize.getWidth();
        }
        return 0;
    }

    public int getHeight() {
        if (viewportSize != null) {
            return viewportSize.getHeight();
        }
        return 0;
    }

    public RectangleSize getViewportSize() {
        return viewportSize;
    }

    public BrowserType getBrowserType() {
        return this.browserType;
    }

    public String getPlatform() {
        if (iosDeviceInfo != null) {
            return "ios";
        }

        if (androidDeviceInfo != null) {
            return "android";
        }

        if (browserType != null) {
            switch (this.browserType) {
                case CHROME:
                case CHROME_ONE_VERSION_BACK:
                case CHROME_TWO_VERSIONS_BACK:
                case FIREFOX:
                case FIREFOX_ONE_VERSION_BACK:
                case FIREFOX_TWO_VERSIONS_BACK:
                    return "linux";
                case SAFARI:
                case SAFARI_ONE_VERSION_BACK:
                case SAFARI_TWO_VERSIONS_BACK:
                    return "mac os x";
                case IE_10:
                case IE_11:
                case EDGE:
                case EDGE_LEGACY:
                case EDGE_CHROMIUM:
                case EDGE_CHROMIUM_ONE_VERSION_BACK:
                    return "windows";
            }
        }
        return "linux";
    }

    public EmulationBaseInfo getEmulationInfo() {
        return emulationInfo;
    }

    public IosDeviceInfo getIosDeviceInfo() {
        return iosDeviceInfo;
    }

    public AndroidDeviceInfo getAndroidDeviceInfo() {
        return androidDeviceInfo;
    }

    public String getSizeMode() {
        return this.sizeMode;
    }

    public void setViewportSize(RectangleSize viewportSize) {
        this.viewportSize = viewportSize;
    }

    public String getBaselineEnvName() {
        return baselineEnvName;
    }

    public String getDeviceVersion() {
        return deviceVersion;
    }

    @Override
    public String toString() {
        return "RenderBrowserInfo{" +
                "viewportSize=" + viewportSize +
                ", browserType=" + browserType +
                ", platform='" + platform + '\'' +
                ", emulationInfo=" + emulationInfo +
                ", iosDeviceInfo=" + iosDeviceInfo +
                ", sizeMode='" + sizeMode + '\'' +
                ", baselineEnvName='" + baselineEnvName + '\'' +
                ", androidDeviceInfo=" + androidDeviceInfo +
                '}';
    }
}

File: dotnet/Eyes.Images.DotNet/VisualGrid/RenderBrowserInfo.cs
using System.ComponentModel;
using Applitools.Utils.Geometry;
using Applitools.VisualGrid.Model;
using System.Drawing;
using Newtonsoft.Json;
using BrowserTypes = Applitools.BrowserType;

namespace Applitools.VisualGrid
{
    public class RenderBrowserInfo
    {
        public RenderBrowserInfo(DesktopBrowserInfo desktopBrowserInfo, string baselineEnvName = null)
        {
            DesktopBrowserInfo = desktopBrowserInfo;
            BaselineEnvName = baselineEnvName;
            BrowserType = desktopBrowserInfo.BrowserType;
            ViewportSize = new RectangleSize(desktopBrowserInfo.ViewportSize);
        }

        public RenderBrowserInfo(IosDeviceInfo deviceInfo, string baselineEnvName = null)
        {
            IosDeviceInfo = deviceInfo;
            BaselineEnvName = baselineEnvName;
            BrowserType = BrowserTypes.SAFARI;
        }

        public RenderBrowserInfo(ChromeEmulationInfo chromeEmulationInfo, string baselineEnvName = null)
        {
            ChromeEmulationInfo = chromeEmulationInfo;
            BaselineEnvName = baselineEnvName;
        }

        public RenderBrowserInfo(AndroidDeviceInfo androidDeviceInfo, string baselineEnvName = null)
        {
            AndroidDeviceInfo = androidDeviceInfo;
            BaselineEnvName = baselineEnvName;
        }

        public RectangleSize GetDeviceSize()
        {
            return ViewportSize ??
                   ChromeEmulationInfo?.Size ??
                   IosDeviceInfo?.Size ??
                   Size.Empty;
        }

        public string GetDeviceName()
        {
            return ChromeEmulationInfo?.DeviceName.ToString() ??
                   IosDeviceInfo?.DeviceName.ToString() ??
                   AndroidDeviceInfo?.DeviceName.ToString() ??
                   string.Empty;
        }
        
        public void SetEmulationDeviceSize(DeviceSize size)
        {
            if (size != null && ChromeEmulationInfo != null)
            {
                if (ChromeEmulationInfo.ScreenOrientation == ScreenOrientation.Portrait)
                {
                    ChromeEmulationInfo.Size = size.Portrait;
                }
                else
                {
                    ChromeEmulationInfo.Size = size.Landscape;
                }
            }
        }

        public void SetIosDeviceSize(DeviceSize size)
        {
            if (size != null && IosDeviceInfo != null)
            {
                if (IosDeviceInfo.ScreenOrientation == ScreenOrientation.Portrait)
                {
                    IosDeviceInfo.Size = size.Portrait;
                }
                else
                {
                    IosDeviceInfo.Size = size.Landscape;
                }
            }
        }

        [JsonIgnore]
        public DesktopBrowserInfo DesktopBrowserInfo { get; }
        public ChromeEmulationInfo ChromeEmulationInfo { get; }
        public IosDeviceInfo IosDeviceInfo { get; }
        public AndroidDeviceInfo AndroidDeviceInfo { get; }

        public string BaselineEnvName { get; }
        
        [DefaultValue(SizeMode.FullPage)]
        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
        public SizeMode Target { get; } = SizeMode.FullPage;
        
        [JsonProperty("name")]
        public BrowserType? BrowserType { get; }

        public int? Width => ViewportSize?.Width;
        public int? Height => ViewportSize?.Height;
        [JsonIgnore]
        public RectangleSize ViewportSize { get; set; }
        
        public EnvironmentRenderer Renderer { get; set; }
        
        [JsonIgnore]
        public string Platform
        {
            get
            {
                if (IosDeviceInfo != null) return "ios";
                if (DesktopBrowserInfo != null)
                {
                    switch (DesktopBrowserInfo.BrowserType)
                    {
                        case BrowserTypes.IE_10:
                        case BrowserTypes.IE_11:
#pragma warning disable CS0618
                        case BrowserTypes.EDGE:
#pragma warning restore CS0618
                        case BrowserTypes.EDGE_LEGACY:
                        case BrowserTypes.EDGE_CHROMIUM:
                        case BrowserTypes.EDGE_CHROMIUM_ONE_VERSION_BACK:
                        case BrowserTypes.EDGE_CHROMIUM_TWO_VERSIONS_BACK:
                            return "windows";
                        case BrowserTypes.SAFARI:
                        case BrowserTypes.SAFARI_ONE_VERSION_BACK:
                        case BrowserTypes.SAFARI_TWO_VERSIONS_BACK:
                        case BrowserTypes.SAFARI_EARLY_ACCESS:
                            return "mac os x";
                    }
                }
                return "linux";
            }
        }

        public override string ToString()
        {
            return (DesktopBrowserInfo ?? IosDeviceInfo ?? (IRenderBrowserInfo)ChromeEmulationInfo).ToString();
        }
    }
}
